potassium/potassium.py

```diff
@@ -132,69 +135,81 @@ def test_client(self):
-                    res = make_response(out.json)
-                    res.status_code = out.status
-                    res.headers['X-Endpoint-Type'] = endpoint.type
-                    return res
-                except:
-                    tb_str = traceback.format_exc()
-                    print(colored(tb_str, "red"))
-                    res = make_response(tb_str)
-                    res.status_code = 500
-                    res.headers['X-Endpoint-Type'] = endpoint.type
-                    return res
-
-        if endpoint.type == "background":
+            try:
+                out = endpoint.func(req)
+                res = make_response(out.json)
+                res.status_code = out.status
+                res.headers['X-Endpoint-Type'] = endpoint.type
+            except:
+                tb_str = traceback.format_exc()
+                print(colored(tb_str, "red"))
+                res = make_response(tb_str)
+                res.status_code = 500
+                res.headers['X-Endpoint-Type'] = endpoint.type
+            self._gpu_lock.release()
+        elif endpoint.type == "background":
             req = Request(
                 json=flask_request.get_json()
             )
 
             # run as threaded task
             def task(endpoint, lock, req):
-                with lock:
-                    try:
-                        endpoint.func(req)
-                    except Exception as e:
-                        # do any cleanup before re-raising user error
-                        self._write_event_chan(True)
-                        raise e
-                self._write_event_chan(True)
-
-            thread = Thread(target=task, args=(endpoint, self._lock, req))
+                try:
+                    endpoint.func(req)
+                except Exception as e:
+                    # do any cleanup before re-raising user error
+                    raise e
+                finally:
+                    self._background_task_cv.notify_all()
+                    # release lock
+                    lock.release()
+
+            thread = Thread(target=task, args=(endpoint, self._gpu_lock, req))
             thread.start()
 
             # send task start success message
             res = make_response({'started': True})
             res.headers['X-Endpoint-Type'] = endpoint.type
-            return res
+        else:
+            raise InvalidEndpointTypeException()
 
-        raise InvalidEndpointTypeException()
-
-    def _write_event_chan(self, item):
-        try:
-            self._event_chan.put(item, block=False)
-        except Full:
-            pass
+        return res
 
+    # TODO - cover depends on this being called so it should not be private
     def _read_event_chan(self) -> bool:
-        return self._event_chan.get()
+        """
+        _read_event_chan essentially waits for a background task to finish, 
+        and then returns True
+        """
+        with self._gpu_lock:
```

Comment:
If I understand correctly, this will lock up the potassium app after the use of handler

step 1: wait to acquire gpu lock. This can happen after both handler and background are done. 
step 2: while holding gpu lock, have some blocking wait for a signal that only ever comes from a background.

meaning if that signal never comes (the case of a handler), we're now indefinitely holding the gpu lock, taking down the potassium instance. This wouldn't be caught on normal potassium testing either because it's unique to the banana environment